pkg/template/controller/templateinstance_controller.go

```diff
@@ -132,38 +140,131 @@ func (c *TemplateInstanceController) sync(key string) error {
+	if len(templateInstance.Status.Objects) != len(templateInstance.Spec.Template.Objects) {
+		err = c.instantiate(templateInstance)
+		if err != nil {
+			glog.V(4).Infof("TemplateInstance controller: instantiate %s returned %v", key, err)
+
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:    templateapi.TemplateInstanceInstantiateFailure,
+				Status:  kapi.ConditionTrue,
+				Reason:  "Failed",
+				Message: err.Error(),
+			})
+		}
 	}
 
-	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceReady)
-	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceInstantiateFailure)
-
-	if instantiateErr == nil {
-		templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
-			Type:               templateapi.TemplateInstanceReady,
-			Status:             kapi.ConditionTrue,
-			LastTransitionTime: metav1.Now(),
-			Reason:             "Created",
-		})
-
-	} else {
-		templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
-			Type:               templateapi.TemplateInstanceInstantiateFailure,
-			Status:             kapi.ConditionTrue,
-			LastTransitionTime: metav1.Now(),
-			Reason:             "Failed",
-			Message:            instantiateErr.Error(),
-		})
+	if !templateInstance.HasCondition(templateapi.TemplateInstanceInstantiateFailure, kapi.ConditionTrue) {
+		ready, err := c.checkReadiness(templateInstance)
+		if err != nil {
+			glog.V(4).Infof("TemplateInstance controller: checkReadiness %s returned %v", key, err)
+
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:    templateapi.TemplateInstanceInstantiateFailure,
+				Status:  kapi.ConditionTrue,
+				Reason:  "Failed",
+				Message: err.Error(),
+			})
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:   templateapi.TemplateInstanceReady,
+				Status: kapi.ConditionFalse,
+			})
+
+		} else if ready {
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:   templateapi.TemplateInstanceReady,
+				Status: kapi.ConditionTrue,
+				Reason: "Created",
+			})
+
+		} else {
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:    templateapi.TemplateInstanceReady,
+				Status:  kapi.ConditionFalse,
+				Reason:  "Waiting",
+				Message: "Waiting for instantiated objects to report ready",
+			})
+		}
 	}
 
 	_, err = c.templateclient.TemplateInstances(templateInstance.Namespace).UpdateStatus(templateInstance)
 	if err != nil {
 		utilruntime.HandleError(fmt.Errorf("TemplateInstance status update failed: %v", err))
+		return err
+	}
+
+	if !templateInstance.HasCondition(templateapi.TemplateInstanceReady, kapi.ConditionTrue) {
```

Comment:
shouldn't this be checking if Condition[waiting]==true (or is not set) instead?  Isn't this going to re-queue templateinstances that are in a Failed state?